x/config/configurator.go

```diff
@@ -0,0 +1,230 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package config
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+
+	"go.uber.org/yarpc"
+	"go.uber.org/yarpc/internal/decode"
+	errs "go.uber.org/yarpc/internal/errors"
+
+	"gopkg.in/yaml.v2"
+)
+
+// Configurator helps build Dispatchers using runtime configuration.
+//
+// An empty Configurator does not know about any transports. Inform it about
+// the different transports and their configuration parameters using the
+// RegisterTransport function.
+type Configurator struct {
+	knownTransports map[string]*compiledTransportSpec
+}
+
+// New sets up a new empty Configurator. The returned Configurator does not
+// know about any transports. Individual TransportSpecs must be registered
+// against it using the RegisterTransport function.
+func New() *Configurator {
+	return &Configurator{knownTransports: make(map[string]*compiledTransportSpec)}
+}
+
+// RegisterTransport registers a TransportSpec with the given Configurator. An
+// error is returned if the TransportSpec was invalid.
+//
+// If a transport with the same name was already registered, it will be
+// overwritten.
+//
+// Use MustRegisterTransport if you want to panic in case of registration
+// failure.
+func (c *Configurator) RegisterTransport(t TransportSpec) error {
+	if t.Name == "" {
+		return errors.New("name is required")
+	}
+
+	spec, err := compileTransportSpec(&t)
+	if err != nil {
+		return fmt.Errorf("invalid TransportSpec for %q: %v", t.Name, err)
+	}
+
+	// TODO: Panic if a transport with the given name is already registered?
```

Comment:
This was mostly to solicit opinions on what we should do here. We can,

- Stomp existing values. I like this. It's simple.
- Fail (panic or error) if the name has already been used in which case we'll still want to support replacing old transport specs by providing an `UnregisterTransport` function. I dislike that this introduces another function and makes RegisterTransport flaky in that it's not guaranteed to always work with a valid TransportSpec.